# 铜绿假单胞菌

## 1.数据质控

- 部分数据由华大BGISEQ-500 WGS产出，利用华大自主开发的过滤软件SOAPnuke完成此步分析处理某些原始序列带有adaptor序列，或含有少量低质量序列，具体参数查看report
- 部分数据由诺禾致源Illumina高通量测序平台NovaSeq 6000 产出，使用fastp软件过滤数据，获得的高质量clean data，具体参数查看report

## 2.clean_data的整理

```shell
# 删除掉每个样本目录里的MD5文件
find /home/liuxq/20230423_Pseudomonas_aeruginosa/0-clean_data/2.cleandata-104/ -type d -print0 | while IFS= read -r -d '' dir; do cd "$dir"; rm -f MD5*; done
# 对样本目录名进行修改，只保留下划线左边的字符
for dir in */; do mv -- "$dir" "${dir%%_*}/"; done

find . -maxdepth 1 -type d -name "NY*" -exec bash -c 'cd "{}" && rm -f MD5*' \;
find . -maxdepth 1 -type d -name "NY????" -print -exec bash -c 'cd "{}" && rm -f *{1,2}.fq.gz' \;
for dir in *FDSW*/; do mv -- "$dir" "${dir%%_*}/"; done
for fqgz_file in *_*.fq.gz; do dir_name=$(echo "$fqgz_file" | cut -d'_' -f1); mkdir -p "$dir_name"; mv "$fqgz_file" "$dir_name/"; done
for dir in *A/; do new_dir=$(echo "$dir" | sed 's/A\///'); mv "$dir" "$new_dir"; done

# fq.gz文件名修改为clean.fq.gz
for dir in */; do     cd "$dir"     for fqgz_file in *.fq.gz; do         mv "$fqgz_file" "$(echo "$fqgz_file" | sed 's/\.fq\.gz$/.clean&/')";     done     cd ..; done

# 删除罗新华和陈方舟（非301来源的）51株菌
sed -i 's/\r//' sample_id.txt # 改成Unix文件格式
ls -d * | grep -v -f ../sample_id.txt | xargs rm -rf
```

## 3.fastqc

```sh
conda install -c bioconda fastqc

cd /home/liuxq/20230423_Pseudomonas_aeruginosa/0-clean_data/
for subdir in NY7*; do files=$(cd "$subdir" && ls *fq.gz); file_list=($files); echo "mkdir /home/liuxq/20230423_Pseudomonas_aeruginosa/fastqc/$subdir" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/fastqc/fastqc.sh; echo "fastqc -t 12 -o /home/liuxq/20230423_Pseudomonas_aeruginosa/fastqc/$subdir /home/liuxq/20230423_Pseudomonas_aeruginosa/0-clean_data/$subdir/${file_list[0]} /home/liuxq/20230423_Pseudomonas_aeruginosa/0-clean_data/$subdir/${file_list[1]}" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/fastqc/fastqc.sh; done

nohup bash fastqc.sh > fastqc.log 2>&1 &
```

## 4.trimmomatic

 丢弃平均碱基质量值低于30的reads

```shell
conda install -c bioconda trimmomatic

cd /home/liuxq/20230423_Pseudomonas_aeruginosa/0-clean_data/
for subdir in NY7*; do files=$(cd "$subdir" && ls *fq.gz); file_list=($files); echo "mkdir /home/liuxq/20230423_Pseudomonas_aeruginosa/trimmomatic/$subdir" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/trimmomatic/trimmomatic.sh; echo "trimmomatic PE -threads 36 /home/liuxq/20230423_Pseudomonas_aeruginosa/0-clean_data/$subdir/${file_list[0]} /home/liuxq/20230423_Pseudomonas_aeruginosa/0-clean_data/$subdir/${file_list[1]} -baseout /home/liuxq/20230423_Pseudomonas_aeruginosa/trimmomatic/$subdir/"$subdir"_trim.fq.gz AVGQUAL:30" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/trimmomatic/trimmomatic.sh; echo "rm /home/liuxq/20230423_Pseudomonas_aeruginosa/trimmomatic/$subdir/*U.fq.gz">> /home/liuxq/20230423_Pseudomonas_aeruginosa/trimmomatic/trimmomatic.sh ;done

nohup bash trimmomatic.sh > trim.log 2>&1 &
```

统计trim.log中的信息

```python
f=open("trim.log")
o=open("trim_statistic.txt","w")
for line in f:
	if line.startswith(" -threads"):
		o.write(line.strip().split("/")[5]+"\t")
	if line.startswith("Quality encoding"):
		o.write(line.strip().split(" ")[4]+"\t")
	if line.startswith("Input Read Pairs"):
		o.write(line.strip().split(" ")[3]+"\t"+line.strip().split(" ")[6]+"\t"+line.strip().split(" ")[11]+"\t"+line.strip().split(" ")[16]+"\t"+line.strip().split(" ")[19]+"\n")
```

## 5.genome assembly

```shell
# 安装conda
wget https://mirrors.tuna.tsinghua.edu.cn/anaconda/miniconda/Miniconda3-latest-Linux-x86_64.sh --no-check-certificate
chmod +x Miniconda3-latest-Linux-x86_64.sh
./Miniconda3-latest-Linux-x86_64.sh

# 修改镜像

vim ~/.condarc
channels:
  - http://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main/
  - http://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free/
  - http://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge/
  - http://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/bioconda/
  - http://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/Paddle/
  - http://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/fastai/
  - http://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch/
show_channel_urls: true
ssl_verify: false

# 安装shovill
conda install shovill

############################ bash shovill.sh #############################
#!/bin/bash
cd /home/liuxq/20230423_Pseudomonas_aeruginosa/0-clean_data/
for subdir in *; do files=$(cd "$subdir" && ls *fq.gz); file_list=($files); echo "shovill --outdir /home/liuxq/20230423_Pseudomonas_aeruginosa/3-shovill/$subdir --R1 /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/$subdir/${file_list[0]} --R2 /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/$subdir/${file_list[1]} --gsize 6.26M --ram 16 --cpus 8" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/script/shovill_assembly.sh; done
#########################################################################
split -l 320 -d shovill_assembly.sh shovill_split_
for file in shovill_split_*; do mv $file "$file.sh"; done
for file in shovill_split_*; do nohup bash $file > "$file.log" 2>&1 & done

# 安装spades
wget http://cab.spbu.ru/files/release3.15.5/SPAdes-3.15.5-Linux.tar.gz
tar -zxf SPAdes-3.15.5-Linux.tar.gz
echo 'export PATH=/home/liuxq/software/SPAdes-3.15.5-Linux/bin:$PATH' >> ~/.bashrc
source ~/.bashrc

############################ bash spades.sh ###############################
#!/bin/bash
cd /home/liuxq/20230423_Pseudomonas_aeruginosa/0-clean_data/
for subdir in *; do
  files=$(cd "$subdir" && ls *fq.gz)
  file_list=($files)
  echo "spades.py -1 "/home/liuxq/20230423_Pseudomonas_aeruginosa/0-clean_data/$subdir/${file_list[0]}" -2 "/home/liuxq/20230423_Pseudomonas_aeruginosa/0-clean_data/$subdir/${file_list[1]}" -o "/home/liuxq/20230423_Pseudomonas_aeruginosa/1-spades_assembly/$subdir"" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/script/spades_assembly.sh 
done
############################################################################
split -l 160 -d spades_assembly.sh spades_split_
for file in spades_split_*; do mv $file "$file.sh"; done
for file in spades_split_*; do nohup bash $file > "$file.log" 2>&1 & done
```

## 6.checkm评估基因组的质量

```shell
conda create -n checkm python=3.9
conda activate checkm
conda install -c conda-forge -c bioconda checkm-genome
wget https://data.ace.uq.edu.au/public/CheckM_databases/checkm_data_2015_01_16.tar.gz
tar -xzvf checkm_data_2015_01_16.tar.gz
export CHECKM_DATA_PATH=/workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/1-checkm/my_checkm_data/
checkm data setRoot /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/1-checkm/my_checkm_data/

nohup checkm lineage_wf -t 16 --pplacer_threads 16 --tab_table -f checkm.txt -x fa /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/2-1288_contig output &
```

## 7.菌种鉴定

```powershell
# 提取shovill组装后的contig
for dir in */; do cd "$dir"; cp ./contigs.fa "/home/liuxq/20230423_Pseudomonas_aeruginosa/1-shovill_assembly/1-contig"; mv "/home/liuxq/20230423_Pseudomonas_aeruginosa/1-shovill_assembly/1-contig/contigs.fa" "/home/liuxq/20230423_Pseudomonas_aeruginosa/1-shovill_assembly/1-contig/${dir%/}.fa"; cd ..; done

# 基因组大小
find . -type f -size -5M -exec ls -lh {} \;
find . -type f -size +7M -exec ls -lh {} \;
ls -lh | awk '{print substr($9, 1, index($9,".")-1),$5}' > /home/liuxq/20230423_Pseudomonas_aeruginosa/1-shovill_assembly/contig_size.txt
```

```shell
############################## 计算平均核苷酸一致性 ################################
# 安装fastANI
unzip fastANI-Linux64-v1.33.zip
echo 'export PATH=/home/liuxq/software/:$PATH' >> ~/.bashrc
source ~/.bashrc

cd /home/liuxq/20230423_Pseudomonas_aeruginosa/1-shovill_assembly/1-contig
find "$(pwd)" -type f > /home/liuxq/20230423_Pseudomonas_aeruginosa/2-fastANI/query_files.txt

# reference:PAO1
# No ANI output is reported for a genome pair if ANI value is much below 80%.
# OUTPUT_FILE will contain tab delimited row(s) with query genome, reference genome, ANI value, count of bidirectional fragment mappings, and total query fragments. 
nohup fastANI --ql /home/liuxq/20230423_Pseudomonas_aeruginosa/2-fastANI/query_files.txt -r /home/liuxq/20230423_Pseudomonas_aeruginosa/0-reference/GCF_000006765.1_ASM676v1_genomic.fna -o /home/liuxq/20230423_Pseudomonas_aeruginosa/2-fastANI/PAO1_ani_output.txt -t 48 > /home/liuxq/20230423_Pseudomonas_aeruginosa/2-fastANI/PAO1_ani.log 2>&1 &

awk -F '\t' '{split($1,a,"/"); split(a[length(a)],b,"."); print b[1], $3}' PAO1_ani_output.txt > PAO1_strain_ani.txt

# reference:PA7
nohup fastANI --ql /home/liuxq/20230423_Pseudomonas_aeruginosa/2-fastANI/query_files.txt -r /home/liuxq/20230423_Pseudomonas_aeruginosa/0-reference/GCF_000017205.1_ASM1720v1_genomic.fna -o /home/liuxq/20230423_Pseudomonas_aeruginosa/2-fastANI/PA7_ani_output.txt -t 48 > /home/liuxq/20230423_Pseudomonas_aeruginosa/2-fastANI/PA7_ani.log 2>&1 &

awk -F '\t' '{split($1,a,"/"); split(a[length(a)],b,"."); print b[1], $3}' PA7_ani_output.txt > PA7_strain_ani.txt
```

```shell
############################## 鉴定物种 ####################################
# 安装kraken2(https://github.com/DerrickWood/kraken2,下载并解压缩)
cd /home/liuxq/software/kraken2-master
bash ./install_kraken2.sh /home/liuxq/software/kraken2
vim ~/.bashrc
export PATH=/home/liuxq/software/kraken2/:$PATH
source ~/.bashrc

# 修改/home/liuxq/software/kraken2/rsync_from_ncbi.pl中第120行为以下内容，来解决ncbi中部分文件不存在的问题
else {


 system("rsync --dry-run --no-motd --files-from=manifest.txt rsync://${SERVER}${SERVER_PATH} . 2> rsync.err");
 open ERR_FILE, "<", "rsync.err"
   or die "$PROG: can't read rsync.err file: $!\n";
 while (<ERR_FILE>) {
   chomp;
   # I really doubt this will work across every version of rsync. :(
   if (/failed: No such file or directory/ && /^rsync: link_stat "\/([^"]+)"/) {
     delete $manifest{$1};
   }
 }
 close ERR_FILE;
 print STDERR "Rsync dry run complete, removing any non-existent files from manifest.\n";

 # Rewrite manifest
 open MANIFEST, ">", "manifest.txt"
   or die "$PROG: can't write manifest: $!\n";
 print MANIFEST "$_\n" for keys %manifest;
 close MANIFEST;

 print STDERR "Step 1/2: Performing rsync file transfer of requested files\n";
 system("rsync --no-motd --files-from=manifest.txt rsync://${SERVER}${SERVER_PATH}/ .") == 0
   or die "$PROG: rsync error, exiting: $?\n";
 print STDERR "Rsync file transfer complete.\n";
}

# 构建数据库
nohup kraken2-build --standard --threads 12 --db /home/liuxq/20230423_Pseudomonas_aeruginosa/3-kraken2/0-ref_db > /home/liuxq/20230423_Pseudomonas_aeruginosa/3-kraken2/kraken2_build.log 2>&1 &

# 序列分类
cd /home/liuxq/20230423_Pseudomonas_aeruginosa/1-shovill_assembly/1-contig
for dir in *; do echo "kraken2 --db /home/liuxq/20230423_Pseudomonas_aeruginosa/3-kraken2/0-ref_db --threads 256 /home/liuxq/20230423_Pseudomonas_aeruginosa/1-shovill_assembly/1-contig/$dir --output /home/liuxq/20230423_Pseudomonas_aeruginosa/3-kraken2/1-Classification_stdout/${dir%%.*}.txt --report /home/liuxq/20230423_Pseudomonas_aeruginosa/3-kraken2/1-Classification_report/${dir%%.*}.report" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/3-kraken2/kraken2_classification.sh; done

nohup bash kraken2_classification.sh > kraken2_classification.log 2>&1 &

# 安装Bracken（https://github.com/jenniferlu717/Bracken）
cd /home/liuxq/software/Bracken-master
bash install_bracken.sh
vim ~/.bashrc
export PATH=/home/liuxq/software/Bracken-master/:$PATH
source ~/.bashrc

# 生成bracken数据库文件
nohup bracken-build -d /home/liuxq/20230423_Pseudomonas_aeruginosa/3-kraken2/0-ref_db -t 48 -l 150 > bracken_build.log 2>&1 &

# 物种丰度估计
cd /home/liuxq/20230423_Pseudomonas_aeruginosa/3-kraken2/1-Classification_report
for dir in *; do echo "bracken -d /home/liuxq/20230423_Pseudomonas_aeruginosa/3-kraken2/0-ref_db -i /home/liuxq/20230423_Pseudomonas_aeruginosa/3-kraken2/1-Classification_report/$dir -o /home/liuxq/20230423_Pseudomonas_aeruginosa/4-bracken/Abundance_Estimation/${dir%%.*}.bracken -r 150" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/4-bracken/bracken.sh ; done

nohup bash bracken.sh > bracken.log 2>&1 & 
```

## 8.bowtie2比对

```sh
# 创建 bowtie2 索引文件
bowtie2-build --threads 16 /home/liuxq/20230423_Pseudomonas_aeruginosa/0-reference/GCF_000006765.1_ASM676v1_genomic.fna PAO1
```

```python
# 比对并提取uniq mapping的reads，计算depth
f=open("/home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/sample_qual.txt")
d={}
for line in f:
	t=line.strip().split("\t")
	d[t[0]] = t[1]
o=open("/home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/bowtie2.sh","w")
import os
path="/home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/"
files=os.listdir(path)
for file in files:
	o.write("bowtie2 --"+d[file]+" -p 8"+" -x /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/ref/PAO1 -1 "+path+file+"/"+file+"_trim_1P.fq.gz"+" -2 "+path+file+"/"+file+"_trim_2P.fq.gz"+" -S /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/sam/"+file+".sam"+"\n")
	o.write("samtools view -hS -F 4 -q 20 /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/sam/"+file+".sam | grep -E \"@|AS:i\" | grep -v \"XS:\" > /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/sam/"+file+"_uniq.sam\n")
	o.write("samtools view -bS /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/sam/"+file+"_uniq.sam > /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/sam/"+file+"_uniq.bam\n")
	o.write("samtools sort -O bam -o /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/sam/"+file+"_uniq_sort.bam /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/sam/"+file+"_uniq.bam\n")
	o.write("samtools depth -a /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/sam/"+file+"_uniq_sort.bam > /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/depth/"+file+".depth\n")
	o.write("rm /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/sam/"+file+".sam\n")
	o.write("rm /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/sam/"+file+"_uniq.sam\n")
	o.write("rm /home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/sam/"+file+"_uniq.bam\n")
```

```sh
split -l 1560 -d bowtie2.sh bowtie2_split_
for file in bowtie2_split_*; do mv $file "$file.sh"; done
for file in bowtie2_split_*; do nohup bash $file > "$file.log" 2>&1 & done
```

```python
# 统计测序depth超过10X的碱基占参考基因组的百分比
import os
path="/home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/depth/"
files=os.listdir(path)
o=open("/home/liuxq/20230423_Pseudomonas_aeruginosa/bowtie2/cov10x.txt","w")
for file in files:
	f=open(path+file)
	i=0
	n=0
	for line in f:
		i=i+1
		if int(line.strip().split("\t")[2])>=10:
			n=n+1
	print(i)
	o.write(file.split(".")[0]+"\t"+str(n/i)+"\n")
```

## 9.菌株筛选

> 纳入301医院菌株（1285株）：出发菌株1582株，剔除297株
>
> - [x] 与reference:PAO1的ANI值>=95或者reference:PA7的ANI值>=95（剔除257株非铜绿菌，得到1325株菌）
> - [x] 组装后contig大小在5.5-7.5Mb（剔除有杂菌污染34株，得到1291株菌）
> - [x] 剔除一株不属于301来源的以及一株测序结果不明确的（得到1289株菌）
> - [x] checkm完整度大于95%且污染度小于5%的菌株（剔除4株菌）

> 纳入周老师之前收集（971株）：出发菌株1095株，剔除124株
>
> - [x] 剔除有杂菌污染83株
> - [x] 剔除24株非铜绿菌
> - [x] 剔除1株污染度>5% 
> - [x] 剔除16株环境/动物来源

> **纳入The Pseudomonas Genome** **Database**（5391株）：出发菌株5512株，剔除121株
>
> - [x] 剔除54株完整度<95%或污染度>5%
> - [x] 剔除杂菌污染5株
> - [x] 剔除1株非铜绿
> - [x] 剔除61株与自有数据重复的菌株

```R
# 257株非铜绿的鉴定结果
library(ggplot2)

# 构造数据
data = read.table("257.txt",header = T,sep = "\t")

p=ggplot(data, aes(x = Number, y = reorder(Species, Number), fill = I("#982b2b"))) +
  geom_bar(stat = "identity", width = 0.8) +
  scale_fill_identity() +
  labs(x = "Number", y = "Species") +
  theme_bw()+
  theme(text = element_text(color = "black"),axis.title.x = element_text(size = 12), axis.title.y = element_text(size = 12))
pdf("Distribution of non-Pseudomonas aeruginosa.pdf")
p
dev.off()
```

```Python
# 对batch entrez的结果进行信息提取
f=open("2132_biosample_info.txt",encoding="utf-8")
o=open("2132_info_out.csv","w")
o.write("Biosample accession\thost\tgeographic location\tcollection date\n")
for line in f:
    if line.startswith("Attributes"):
        dic={}
        dic["Biosample accession"]=""
        dic["host"] = ""
        dic["geographic location"] = ""
        dic["collection date"] = ""
    if "/host=" in line:
        dic["host"] = line.strip().split("=")[1].replace('"', '')
    if "/geographic location=" in line:
        dic["geographic location"] = line.strip().split("=")[1].replace('"', '')
    if "/collection date=" in line:
        dic["collection date"] = line.strip().split("=")[1].replace('"', '')
    if line.startswith("Accession:"):
        dic["Biosample accession"] = line.strip().split("\t")[0].split(" ")[1]
        o.write(dic["Biosample accession"]+"\t"+dic["host"]+"\t"+dic["geographic location"]+"\t"+dic["collection date"]+"\n")
```

## 10.variant calling

```python
# 将筛选后的1289株中的菌株移动到另一个目录
import os
path = "/home/liuxq/20230423_Pseudomonas_aeruginosa/7-bowtie2/sam/"
files=os.listdir(path)
f=open("/home/liuxq/20230423_Pseudomonas_aeruginosa/7-bowtie2/1283_strain.txt")
o=open("/home/liuxq/20230423_Pseudomonas_aeruginosa/7-bowtie2/mv.sh","w")
d=[]
for line in f:
	d.append(line.strip())
for file in files:
	if file.strip().split("_")[0] in d:
		o.write("mv "+path+file+" /home/liuxq/20230423_Pseudomonas_aeruginosa/7-bowtie2/bam\n")
```

```shell
cd /home/liuxq/20230423_Pseudomonas_aeruginosa/7-bowtie2/bam

for file in *; do prefix=$(echo $file | cut -d "_" -f 1); echo "samtools mpileup -q 1 -f /home/liuxq/20230423_Pseudomonas_aeruginosa/0-reference/GCF_000006765.1_ASM676v1_genomic.fna /home/liuxq/20230423_Pseudomonas_aeruginosa/7-bowtie2/bam/$file 1>/home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/mpileup/$prefix.mpileup 2>/home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/log/$prefix.mpileup.log" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/call_variant.sh; echo "varscan mpileup2snp /home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/mpileup/$prefix.mpileup --output-vcf 1 > /home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/snp/$prefix.varscan.snp.vcf 2>/home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/log/$prefix.varscan.snp.vcf.log" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/call_variant.sh; echo "varscan mpileup2indel /home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/mpileup/$prefix.mpileup --output-vcf 1 > /home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/indel/$prefix.varscan.indel.vcf 2>/home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/log/$prefix.varscan.indel.vcf.log" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/call_variant.sh;echo "rm /home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/mpileup/$prefix.mpileup">>/home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling/call_variant.sh  ;done

cd /home/liuxq/20230423_Pseudomonas_aeruginosa/variant_calling
split -l 424 -d call_variant.sh variant_split_
for file in variant_split_*; do mv $file "$file.sh"; done
for file in variant_split_*; do nohup bash $file > "$file.log" 2>&1 & done
```

## 11.ST型鉴定

> Multilocus sequence typing (MLST) is an unambiguous procedure for characterising isolates of bacterial species using the sequences of internal fragments of (usually) seven house-keeping genes. Approximately 450-500 bp internal fragments of each gene are used, as these can be accurately sequenced on both strands using an automated DNA sequencer. For each house-keeping gene, the different sequences present within a bacterial species are assigned as distinct alleles and, for each isolate, the alleles at each of the seven loci define the allelic profile or sequence type (ST). 

```shell
conda install mlst

# 对下载的pubmlst数据库进行更新
which mlst-download_pub_mlst
cd ~/miniconda3/bin
./mlst-download_pub_mlst -d ./pubmlst | bash
mv ../db/pubmlst ../db/pubmlst.old # Save the old database folder
mv pubmlst/ ../db/ # Put the new folder there
./mlst-make_blast_db # Regenerate the BLAST database

# MLST分型
cd /home/liuxq/20230423_Pseudomonas_aeruginosa/3-shovill/1-contig
for file in *; do echo "mlst --threads 36 $file 1>/home/liuxq/20230423_Pseudomonas_aeruginosa/mlst/result/${file%%.*}.txt 2>/home/liuxq/20230423_Pseudomonas_aeruginosa/mlst/log/${file%%.*}.log" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/mlst/mlst.sh; done
nohup bash /home/liuxq/20230423_Pseudomonas_aeruginosa/mlst/mlst.sh > /home/liuxq/20230423_Pseudomonas_aeruginosa/mlst/nohup.out &

cd /home/liuxq/20230423_Pseudomonas_aeruginosa/mlst/result
cat * > /home/liuxq/20230423_Pseudomonas_aeruginosa/mlst/1582_mlst.txt
```

结果中有21株未被鉴定出的ST型，使用`--novel`提取相应的新的等位基因，上传到pubmlst数据库（NY8766和NY12408的mutL基因区域有indel），等待返回新基因的id后上传isolate和new profile获取新的ST型，最终共鉴定出279个ST型

```shell
mlst --threads 20 --novel /home/liuxq/20230423_Pseudomonas_aeruginosa/DATABASE/59_new_alleles/NY11052.fa /home/liuxq/20230423_Pseudomonas_aeruginosa/9-mlst/1-strain/NY11052.fa 1>/home/liuxq/20230423_Pseudomonas_aeruginosa/DATABASE/59_st_novel/NY11052.txt 2>/home/liuxq/20230423_Pseudomonas_aeruginosa/DATABASE/59_st_novel_log/NY11052.log
```

```sh
conda create -n srst2 python=2.7
conda activate srst2
conda install -c bioconda srst2

getmlst.py --species 'Pseudomonas aeruginosa'

while read var; do     files=$(cd "/home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/$var" && ls *fq.gz); file_list=($files);     echo "srst2 --input_pe /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/$var/${file_list[0]} /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/$var/${file_list[1]} --forward _trim_1P --reverse _trim_2P --output /home/liuxq/20230423_Pseudomonas_aeruginosa/10-srst2/result/$var --log --mlst_db /home/liuxq/20230423_Pseudomonas_aeruginosa/10-srst2/db/Pseudomonas_aeruginosa.fasta --mlst_definitions /home/liuxq/20230423_Pseudomonas_aeruginosa/10-srst2/db/profiles_csv --mlst_delimiter _ --threads 8" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/10-srst2/srst2.sh; done < /home/liuxq/20230423_Pseudomonas_aeruginosa/9-mlst/no_ST_sample.txt

split -l 86 -d srst2.sh srst2_split_
for file in srst2_split_*; do mv $file "$file.sh"; done
for file in srst2_split_*; do nohup bash $file > "$file.log" 2>&1 & done
```

## 12.血清型鉴定

> The *Pseudomonas aeruginosa* serotyper (PAst) is a command-line-tool for fast and reliable *in silico* serotyping of *P. aeruginosa* isolates, based on whole genome sequence assembly data. 
>
> [Sandramses/PAst: In silico serotyping of Pseudomonas aeruginosa isolates (github.com)](https://github.com/Sandramses/PAst) 

| Serogroup | Reference OSA cluster | Reference gene | Size (bp) | Serotype(s) within serogroup |
| :-------: | :-------------------: | :------------: | :-------: | :--------------------------: |
|    O1     |          O1           |                |  18,368   |              O1              |
|    O2     |          O2           |     *wzyβ*     |  23,303   |           O2, O16            |
|    O3     |          O3           |                |  20,210   |           O3, O15            |
|    O5     |          O2           |                |  23,303   |         O5, O18, O20         |
|    O4     |          O4           |                |  15,279   |              O4              |
|    O6     |          O6           |                |  15,649   |              O6              |
|    O7     |          O7           |                |  19,617   |            O7, O8            |
|    O9     |          O9           |                |  17,263   |              O9              |
|    O10    |          O10          |                |  17,635   |           O10, O19           |
|    O11    |          O11          |                |  13,868   |           O11, O17           |
|    O12    |          O12          |                |  25,864   |             O12              |
|    O13    |          O13          |                |  14,316   |           O13, O14           |

```sh
USAGE

Usage: "./PAst.pl path/to/BLASTbin path/to/output/directory path/to/input/directory path/to/OSAdb"

BLAST bin: Path to the bin containing blasts executables (ex. /usr/bin/ncbi-blast-2.2.29+/bin/)
Output dir: Directory where blast reports, OSA multifasta files and result summary (serotyping.txt) are placed
Input dir: Directory containing assembled input genomes in multifasta format (and nothing else)
OSA database: Path to OSA database file (downloadable from Github)

nohup ./PAst.pl ~/miniconda3/envs/past/bin/ /home/liuxq/20230423_Pseudomonas_aeruginosa/12-PAst/Serotype /home/liuxq/20230423_Pseudomonas_aeruginosa/3-shovill/2-1289_contig /home/liuxq/20230423_Pseudomonas_aeruginosa/12-PAst/OSAdb.fasta &
```

coverage of OSA的阈值为大于95，共鉴定出12个血清型，有15株菌不可分型，4株菌为多分型

## 13.克隆复合物的鉴定

> - Clonal complexes were defined as a group of isolates with either identical STs or STs that varied at one or two loci (single or double-locus variants)
> - When different STs shared alleles at five or six out of seven loci, the STs were grouped as a clonal complex (CC).
> - Clonal complexes (CCs) were defined with the software START2 as a group of STs sharing at least 5 loci 

 [goeBURST tutorial (phyloviz.net)](http://www.phyloviz.net/goeburst/Tutorial.html) 

最终共鉴定出28个clonal complex

```R
# CC地域分布堆积柱状图
setwd("C:/Users/liuxi/Desktop")
df=read.table("cc_region.txt",header=T,sep="\t")
library("reshape2")
library(ggplot2)
library(cowplot)
library(ggsci)
library(tidyverse)
data=melt(df,id="region",variable.name="CC",value.name="number")
data$region=factor(data$region,levels=c("Asia","Europe","North America","South America","Oceania","Africa"))
dat2=data %>% group_by(CC) %>% summarise(total_number=sum(number))
p=ggplot(data, aes(x=CC, y=number,fill=region)) + 
  geom_bar(stat="identity",position = "fill",width = 0.8)+
  geom_text(data=dat2,
            aes(x=CC,y=1,
                label=paste0(total_number)),
            inherit.aes = FALSE,
            vjust=-0.2)+
  theme_bw()+
  theme(panel.grid=element_blank(),
        axis.text.x = element_text(size = 12,angle = 30,hjust = 0.75,vjust = 0.85),
        axis.title = element_text(size = 15),
        legend.text = element_text(size = 13))+
  scale_fill_manual(values = pal_npg("nrc",alpha = 0.8)(6))+
  labs(x=NULL,y="Percentage of Strains")+
  guides(fill=guide_legend(title = "Region"))
pdf("CC_region.pdf",width = 15,height=5)
p
dev.off()

# 地域的CC组成图
setwd("C:/Users/liuxi/Desktop")
df=read.table("region_cc.txt",header=T,sep="\t",check.names=F)
library("reshape2")
library(ggplot2)
library(cowplot)
library(ggsci)
library(tidyverse)
data=melt(df,id="CC",variable.name="region",value.name="number")
asia=data[data$region=="Asia",c(1,3)]
d=asia[order(asia$number),]$CC
data$CC=factor(data$CC,levels = c("Others",d[-26]))
data$region=factor(data$region,levels = rev(c("Asia","Europe","North America","South America","Oceania","Africa")))
dat2=data %>% group_by(region) %>% summarise(total_number=sum(number))
p=ggplot(data, aes(x=number, y=region,fill=CC)) + 
  geom_bar(stat="identity",position = "fill",width = 0.6)+
  geom_text(data=dat2,
          aes(x=1+0.03,y=region,
              label=paste0(total_number)),
          inherit.aes = F,
          vjust=0)+
  theme_bw()+
  theme(panel.grid=element_blank(),
        axis.text.y = element_text(size = 13),
        axis.title = element_text(size = 15),
        legend.text = element_text(size = 12))+
  scale_fill_manual(values = rev(c(pal_nejm("default", alpha = 0.8)(8),pal_d3("category10",alpha = 0.8)(10),pal_lancet(palette = c("lanonc"), alpha = 0.8)(8))))+
  labs(x=NULL,y=NULL)+
  guides(fill=guide_legend(title = NULL,reverse = T))
pdf("region_CC.pdf",width = 15,height =5)
p
dev.off()

# CC与血清型的相关性
setwd("C:/Users/liuxi/Desktop")
df=read.table("cc_serotype.txt",header=T,sep="\t",check.names=F)
library("reshape2")
library(ggplot2)
library(cowplot)
library(ggsci)
library(tidyverse)
data=melt(df,id="CC",variable.name="serotype",value.name="number")
dat2=data %>% group_by(CC) %>% summarise(total_number=sum(number))
data$CC=factor(data$CC,levels = dat2[order(dat2$total_number,decreasing =T),]$CC)
data$serotype=factor(data$serotype,levels = c("O1","O2","O3","O4","O5","O6","O7","O9","O10","O11","O12","O13"))
p=ggplot(data, aes(x=CC, y=number,fill=serotype)) + 
  geom_bar(stat="identity",position = "stack",width = 0.8)+
  geom_text(data=dat2,
            aes(x=CC,y=total_number,
                label=paste0(total_number)),
            inherit.aes = FALSE,
            vjust=-0.2)+
  theme_bw()+
  theme(panel.grid=element_blank(),
        axis.text.x = element_text(size = 12,angle = 30,hjust = 0.75,vjust = 0.85),
        axis.title = element_text(size = 15),
        legend.text = element_text(size = 13))+
  scale_fill_manual(values = pal_d3("category20",alpha = 0.8)(12))+
  labs(x=NULL,y="Number of Strains")+
  guides(fill=guide_legend(title = "Serotype"))

pdf("cc_serotype.pdf",width = 12,height = 7)
p
dev.off()

# CC与时间的相关性
setwd("C:/Users/liuxi/Desktop")
df=read.table("CC-time.txt",header=T,sep="\t",check.names=F)
library("reshape2")
library(ggplot2)
library(cowplot)
library(ggsci)
library(tidyverse)
library(ggalluvial)
data=melt(df,id="CC",variable.name="time",value.name="number")
data$CC=factor(data$CC,levels = rev(c("CC235","CC244","CC111","CC357","CC298","CC277","CC233","CC308","CC175","CC654")))
p=ggplot(data, aes(x=time, y=number,fill=CC,stratum=CC,alluvium=CC)) + 
  geom_bar(stat="identity",position = "stack",width = 0.5,color="black")+
  geom_flow(width=0.5, alpha=0.3, knot.pos=0)+
  theme_cowplot()+
  scale_fill_manual(values = pal_npg("nrc")(10))+
  labs(x="Year",y="Number of Strains")+
  guides(fill=guide_legend(title = NULL))+
  theme(axis.text.x = element_text(size = 13),
        axis.text.y = element_text(size = 13),
        axis.title = element_text(size = 15),
        legend.text = element_text(size = 13))

pdf("CC-time.pdf",width = 12,height = 7)
p
dev.off()

setwd("C:/Users/liuxi/Desktop")
df=read.table("cc-time-trend.txt",header=T,sep="\t",check.names=F)
library("reshape2")
library(ggplot2)
library(cowplot)
library(ggsci)
library(tidyverse)
library(ggalluvial)
data=melt(df,id="CC",variable.name="time",value.name="number")
data$CC=factor(data$CC,levels = rev(c("CC235","CC244","CC111","CC357","CC274")))
p=ggplot(data, aes(x=time, y=number,fill=CC,stratum=CC,alluvium=CC)) + 
  geom_col(width = 0.4,color="black")+
  geom_flow(width=1/3, alpha=0.3, knot.pos=1/4)+
  theme_bw()+
  scale_fill_manual(values = c(pal_npg("nrc")(5)))+
  labs(x="Year",y="Number of Strains")+
  guides(fill=guide_legend(title = NULL))+
  theme(panel.grid=element_blank(),
        axis.text.x = element_text(size = 13),
        axis.text.y = element_text(size = 13),
        axis.title = element_text(size = 15),
        legend.text = element_text(size = 13))

pdf("cc-time-trend.pdf",width = 12,height = 7)
p
dev.off()
```

## 14.耐药基因的鉴定

```bash
conda install -c bioconda ncbi-amrfinderplus

# 更新数据库
amrfinder -u

cd /home/liuxq/20230423_Pseudomonas_aeruginosa/3-shovill/2-1288_contig/
for dir in *; do echo "amrfinder -n /home/liuxq/20230423_Pseudomonas_aeruginosa/3-shovill/2-1288_contig/$dir -O Pseudomonas_aeruginosa --plus -o /home/liuxq/20230423_Pseudomonas_aeruginosa/13-amrfinder/result/${dir%%.*}.txt" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/13-amrfinder/amrfinder.sh; done

nohup bash amrfinder.sh &
```

```python
# 提取identity>=94且coverage>=94的结果
import os 
path="/home/liuxq/20230423_Pseudomonas_aeruginosa/13-amrfinder/result/"
o=open("/home/liuxq/20230423_Pseudomonas_aeruginosa/13-amrfinder/amrgene_result.txt","w")
files=os.listdir(path)
for file in files:
	f=open(path+file)
	for line in f:
		id=file.split(".")[0]
		if not line.startswith("Pro"):
			if float(line.split("\t")[15])>=94 and float(line.split("\t")[16])>=94:
				o.write(id+"\t"+line)
```

> CLSI M100 ED33: [Dashboard (edaptivedocs.net)](http://em100.edaptivedocs.net/dashboard.aspx) 

```R
# 药敏试验结果统计
data=read.table("dst.txt",header = T)
library(ggplot2)
library(tidyverse)

data$dst<-factor(data$dst,levels = c("Unknown","Susceptible","Intermediate","Resistant"))
row = data[data$dst=="Resistant",]
sort_row <- row[order(row$Number,decreasing = T),]
data$Antibiotics<-factor(data$Antibiotics,levels = sort_row$Antibiotics)
# 使用ggplot2绘制堆积柱状图
color = c("#7E6148","#3C5488","#00A087","#A52502")
pdf("Antimicrobial_susceptibility.pdf")
ggplot(data, aes(x = Antibiotics, y = Number, fill = dst)) +
  geom_bar(position = "fill",stat="identity") +
  theme(panel.background = element_rect(fill = NA),panel.border = element_rect(fill = NA),axis.text.x = element_text(angle = 45,hjust = 1,size=12),legend.text = element_text(size = 12),axis.title = element_text(size = 15))+
  scale_fill_manual(values=color)+
  scale_y_continuous(labels = scales::percent_format())+
  labs(x = "Antibiotics", y = "Susceptibility",fill = "dst")
dev.off()
```

```R
# ARG class注释bar的生成
library(pheatmap)
library(ggsci)
setwd("C:/Users/liuxi/Desktop")
data=read.csv("aaa.csv",header = T,check.names = F,row.names = 1)
anno = read.csv("annotation.csv",header = T,check.names = F,row.names = 1)
color=pal_d3("category20")(12)
names(color)=unique(anno$ARG)
p=pheatmap(data,cluster_cols = F,fontsize_col=4,annotation_col = anno,annotation_colors = list(ARG = color))

pdf("res.pdf",width = 10)
p
dev.off()
```

```R
library(pheatmap)
library(ggsci)
setwd("C:/Users/liuxi/Desktop")
data=read.table("CC_resistant.txt",header = T,check.names = F,row.names = 1)
pdf("CC_resistant_heatmap.pdf")
p=pheatmap(data,cluster_rows = T,cluster_cols = F,angle_col = 45,cellwidth = 15, cellheight = 12)
dev.off()
```

```R
library(PRECAST)
library(ggpubr)
setwd("C:/Users/liuxi/Desktop")
data=read.table("CC-class.txt",header = T)
data$CC = factor(data$CC,levels = c("HiRiCs","CC316","CC1976","CC309","CC260","Others"))
my_comparisons=list(c("HiRiCs","Others"),
                    c("CC316","Others"),
                    c("CC1976","Others"),
                    c("CC309","Others"),
                    c("CC260","Others")
                     )
p=ggboxplot(data, x = "CC", y = "Class",       
            color = "black",
            fill = "CC",
            #add = "jitter",
            palette = c('lancet'),
            width = 0.6,
            outlier.shape = NA,
            bxp.errorbar=T)+
  labs(x="CC",y="No. of antibiotic classes")+
  theme(legend.position="none",
        axis.text = element_text(size = 15),
        axis.title = element_text(size = 18))+
  stat_compare_means(comparisons = my_comparisons,method = "t.test")
pdf("CC-class-boxplot.pdf",height = 6)
p
dev.off()

library(PRECAST)
library(ggpubr)
setwd("C:/Users/liuxi/Desktop")
data=read.table("CC-number.txt",header = T)
data$CC = factor(data$CC,levels = c("HiRiCs","CC316","CC1976","CC309","CC260","Others"))
my_comparisons=list(c("HiRiCs","Others"),
                    c("CC316","Others"),
                    c("CC1976","Others"),
                    c("CC309","Others"),
                    c("CC260","Others")
)
p=ggboxplot(data, x = "CC", y = "number",fill = "CC",      
            color = "black",
            palette = c('lancet'),
            width = 0.6,
            outlier.shape = NA,
            bxp.errorbar=T)+
  labs(x="CC",y="No. of resistance genes")+
  theme(legend.position="none",
        axis.text = element_text(size = 15),
        axis.title = element_text(size = 18))+
  stat_compare_means(comparisons = my_comparisons,method = "t.test")
pdf("CC-number-boxplot.pdf",height = 8)
p
dev.off()

library(PRECAST)
library(ggpubr)
setwd("C:/Users/liuxi/Desktop")
data=read.table("cc30-number.txt",header = T)
data$CC = factor(data$CC,levels = c("CC235","CC244","CC111","CC357","CC298","CC277","CC233","CC308","CC175","CC654","CC316","CC1976","CC309","CC260","CC179","CC155","CC348","CC252","CC390","CC253","CC395","CC245","CC17","CC27","CC549","CC262","CC146","CC319","CC274","CC241"))
# my_comparisons=list(c("HiRiCs","Others"),
#                     c("CC316","Others"),
#                     c("CC1976","Others"),
#                     c("CC309","Others"),
#                     c("CC260","Others")
# )
p=ggboxplot(data, x = "CC", y = "number",       
            color = "CC",
            #fill = "CC",
            add = "jitter",
            #palette = c('lancet'),
            width = 0.6,
            outlier.shape = NA,
            bxp.errorbar=T)+
  labs(x="CC",y="No. of resistance genes")+
  theme(legend.position="none",
        axis.text.x  = element_text(size = 10,angle = 30,vjust = 0.5),
        axis.title = element_text(size = 18))
  #stat_compare_means(comparisons = my_comparisons,method = "t.test")
pdf("CC-number-boxplot-30.pdf",width = 12)
p
dev.off()


```

```R
setwd("C:/Users/liuxi/Desktop")
df=read.table("mdr-cc.txt",header=T,sep="\t",check.names=F)
library("reshape2")
library(ggplot2)
library(cowplot)
library(ggsci)
library(tidyverse)
data=melt(df,id="resistance",variable.name="CC",value.name="number")
data$resistance=factor(data$resistance,levels = c("MDR","XDR","Other"))
p=ggplot(data, aes(x=CC, y=number,fill=resistance)) + 
  geom_bar(stat="identity",position = "fill",width = 0.8)+
  theme_bw()+
  theme(panel.grid=element_blank(),
        axis.text.x = element_text(size = 13,angle = 30,hjust = 0.75,vjust = 0.85),
        axis.text.y = element_text(size = 15),
        axis.title = element_text(size = 18),
        legend.title  = element_text(size = 18),
        legend.text = element_text(size = 15))+
  scale_fill_manual(values = c("#456990","#EF767A","#48C0AA"))+
  labs(x=NULL,y="Percentage of Strains")+
  guides(fill=guide_legend(title = NULL))

pdf("mdr-cc.pdf",width = 7,height = 7)
p
dev.off()
```



## 15.质粒的鉴定

精细注释质粒的脚本在100服务器：/workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/13-plasmid_identification/script下

>  [PlasmidSeeker](https://github.com/bioinfo-ut/PlasmidSeeker) 

```bash
# PlasmidSeeker
cd /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288
for subdir in *; do files=$(cd "$subdir" && ls *fq.gz); file_list=($files); echo "gunzip -c /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/${file_list[0]} > /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/1.fq" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidseeker/plasmidseeker.sh; echo "gunzip -c /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/${file_list[1]} > /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/2.fq" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidseeker/plasmidseeker.sh; echo "cat /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/1.fq /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/2.fq > /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/merge.fq" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidseeker/plasmidseeker.sh; echo "/home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidseeker/PlasmidSeeker-master/plasmidseeker.pl -i /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/merge.fq -d /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidseeker/db_w20 -b /home/liuxq/20230423_Pseudomonas_aeruginosa/0-reference/GCF_000006765.1_ASM676v1_genomic.fna -o /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidseeker/result/$subdir.txt" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidseeker/plasmidseeker.sh; echo "rm /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/1.fq /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/2.fq /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/merge.fq" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidseeker/plasmidseeker.sh; done
```

>  [kcri-tz/plasmidfinder (github.com)](https://github.com/kcri-tz/plasmidfinder) 

```bash
cd /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288
for subdir in *; do files=$(cd "$subdir" && ls *fq.gz);  file_list=($files); echo "mkdir /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidfinder/result/$subdir" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidfinder/plasmidfinder.sh; echo "mkdir /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidfinder/tmp/$subdir" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidfinder/plasmidfinder.sh; echo "plasmidfinder.py -i /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/${file_list[0]} /home/liuxq/20230423_Pseudomonas_aeruginosa/2-trimmomatic/1288/$subdir/${file_list[1]} -o /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidfinder/result/$subdir -tmp /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidfinder/tmp/$subdir -p /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidfinder/plasmidfinder_db/ -x" >> /home/liuxq/20230423_Pseudomonas_aeruginosa/14-plasmidfinder/plasmidfinder.sh; done
```

## 16.毒力基因

>  [tseemann/abricate: :mag_right: Mass screening of contigs for antimicrobial and virulence genes (github.com)](https://github.com/tseemann/abricate) 

```bash
abricate --list
abricate --setupdb
abricate-get_db --db vfdb --force #一定要更新，序列数量相差很多
nohup abricate --db vfdb /home/liuxq/20230423_Pseudomonas_aeruginosa/3-shovill/2-1288_contig/* --quiet --nopath > vf.txt &
```

```R
setwd("C:/Users/liuxi/Desktop")
library(ggplot2)
library(ComplexHeatmap)
library(reshape2)
library(RColorBrewer)
library(circlize)
df=read.csv("vf-v5.txt",header = T,sep = "\t")
df$value <- rep(1, nrow(df))
mat=dcast(df,class+gene ~ sample)
mat$Gene = paste(mat$gene,mat$class,sep="_")
tree = read.csv("7647_sort_filename.txt",header = F)
data = mat[,c("Gene",rev(tree$V1))]
rownames(data) = data$Gene
data=data[,-1]
plot=as.matrix(data)
#write.table(plot,"mat.txt",quote = F,sep = "\t")
write.table(mat,"vfgene_matrix.txt",quote = F,sep = "\t")
col_anno = HeatmapAnnotation(Function = mat$class,
                             annotation_width = unit(2, "mm"),
                             simple_anno_size_adjust = TRUE,
                             annotation_name_side = "top",
                             which = "row")
pdf("vf-heatmap-3.pdf",width = 10,height = 10)
Heatmap(plot,
        cluster_rows = F,
        cluster_columns = F,
        na_col = "white",
        border = F,
        use_raster = F,
        row_names_side = "right",
        row_names_gp = gpar(fontsize = 2),
        column_names_gp = gpar(fontsize = 0),
        left_annotation = col_anno,
        heatmap_legend_param = list(title = "Virulence Factors",at = c(0,1),color_bar = "discrete"),
        col = colorRamp2(c(0,1),c("#f8f8fd","#90BCD5")))
dev.off()
```

## 17.菌株的时空分布情况

菌株在全球以及全国的分布：

```Python
import pandas as pd  # 做表格操作的模块
from pyecharts.charts import Map , Grid # 绘图的模块
from pyecharts import options as opts
from pyecharts.render import make_snapshot # 导入输出图片工具
from snapshot_selenium import snapshot # 使用snapshot-selenium 渲染图片
# 支持的国家名字https://blog.csdn.net/qq_16543239/article/details/105784635
pieces = [
    {"min": 2500, "color":"#990033"},
    {"min": 1000, "max": 2500},
    {"min": 400, "max": 1000},
    {"min": 200, "max": 400},
    {"min": 150, "max": 200},
    {"min": 100, "max": 150},
    {"min": 50, "max": 100},
    {"min": 40, "max": 50},
    {"min": 30, "max": 40},
    {"min": 20, "max": 30},
    {"min": 15, "max": 20},
    {"min": 10, "max": 15},
    {"min":  5, "max": 10},
    {"min":  0, "max":  5},
]

df = pd.read_csv('country_strain_number.csv')
Country = df['Country'].tolist()
count = df['count'].tolist()
print(Country)
print(count)
#name_map=name_map,
c = (
    Map(init_opts=opts.InitOpts(width='1400px', height='600px',renderer="svg"))
        .add("累计确诊", [list(z) for z in zip(Country, count)], "world",  is_map_symbol_show=False)
        .set_series_opts(label_opts=opts.LabelOpts(is_show=False))
        .set_global_opts(
        title_opts=opts.TitleOpts(title="铜绿菌株全球分布情况"),
        visualmap_opts=opts.VisualMapOpts(max_=1000000, is_piecewise=True, pieces=pieces),
    )
         .render("map_world.html")
)

make_snapshot(snapshot, "map_world.html", "map_world.svg")
```

```Python
# -*- coding: utf-8 -*-
import pandas as pd  # 做表格操作的模块
from pyecharts.charts import Map  # 绘图的模块
from pyecharts import options as opts
from pyecharts.render import make_snapshot # 导入输出图片工具
from snapshot_selenium import snapshot # 使用snapshot-selenium 渲染图片

pieces = [
    {"min": 1000, "max": 1500, "color":"#990033"},
    {"min": 500, "max": 1000},
    {"min": 200, "max": 500},
    {"min": 100, "max": 200},
    {"min": 50, "max": 100},
    {"min": 20, "max": 50},
    {"min": 10, "max": 20},
    {"min": 5, "max": 10},
    {"min": 0, "max": 5},
]

df = pd.read_csv('province_strain_number.csv')
province = df['province'].tolist()
count = df['count'].tolist()
print(province)
print(count)
c = (
    Map(init_opts=opts.InitOpts(width='1400px', height='600px',renderer='svg'))
        .add("累计确诊", [list(z) for z in zip(province, count)], "china", is_map_symbol_show=False)
        .set_series_opts(label_opts=opts.LabelOpts(is_show=False))
        .set_global_opts(
        title_opts=opts.TitleOpts(title="铜绿菌株全国分布情况"),
        visualmap_opts=opts.VisualMapOpts(max_=250, is_piecewise=True, pieces=pieces),
    )
    .render("map_china.html")
)

make_snapshot(snapshot, "map_china.html", "map_china.svg")
```

```R
library(ComplexHeatmap)
library(circlize)
library(RColorBrewer)
setwd("C:/Users/liuxi/Desktop")
df = read.csv("country_time2.csv",row.names = 1)
df = as.matrix(df)
col_anno = HeatmapAnnotation(Region = c(rep("Asia",23),rep("Africa",12),rep("Europe",24),rep("North America",5),rep("Oceania",3),rep("South America",7),"Where"),
      col = list(Region = c("Asia" = "#8DD3C7","Africa"="#FCCDE5","Europe"="#BEBADA","North America"="#FB8072","Oceania"="#80B1D3","South America"="#FDB462","Where"="#B3DE69")),
      annotation_height = unit(2, "mm"),
      simple_anno_size_adjust = TRUE,
      gp = gpar(col = "grey", lwd = 0.5),
      annotation_name_side = "left")
      
pdf("country_time_distribution2.pdf",width = 12,height = 8)
Heatmap(df,
        cluster_rows = F,
        cluster_columns = F,
        na_col = "white",
        border = T,
        rect_gp = gpar(col = "grey", lwd = 0.5),
        row_names_side = "left",
        row_names_gp = gpar(fontsize = 7.5),
        column_names_gp = gpar(fontsize = 9),
        bottom_annotation = col_anno,
        heatmap_legend_param = list(title = "Number",at = c(0,10,25,50,100,300,500,700),color_bar = "discrete"),
        col = colorRamp2(c(0,10,25,50,100,150,200,400,700),c("#385EAF","#5DBED8","#7FAE9F","#E9BE63","#E18660","#DC615E","#D20046","#B5003C","#990033")))
dev.off()
```

```R
library(ggplot2)
setwd("C:/Users/liuxi/Desktop")
# 构造数据
data = read.table("country_number.txt",header = T,sep = "\t",quote = "")
pdf("country_number.pdf",width=10,height = 7)
bar=barplot(height=data$number,border = NA,col="#385EAF",ylab = "Number",ylim=c(0,3000),width=2)
text(x=bar,y=data$number+25,data$number,cex=0.4,font=2)
dev.off()

data = read.table("time_number.txt",header = T,sep = "\t",quote = "")
pdf("time_number.pdf",width=7,height = 10)
bar=barplot(height=data$number,border = NA,col="#385EAF",xlab = "Number",xlim=c(0,1200),width=2,horiz = T)
text(x=data$number+15,y=bar,data$number,cex=0.4,font=2)
dev.off()
```

## 18.基因组注释与比对

```shell
conda create -n prokka_env -c conda-forge -c bioconda prokka
prokka --setupdb
```

```python
o=open("./prokka.sh","a")
import os
path="/workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/2-1285_contig/"
files=os.listdir(path)
for file in files:
	o.write("prokka "+path+file+" --outdir /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/4-prokka/result --prefix "+file.split(".f")[0]+" --force --centre X --compliant\n")
	o.write("rm /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/4-prokka/result/"+file.split(".f")[0]+".err\n")
	o.write("rm /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/4-prokka/result/"+file.split(".f")[0]+".faa\n")
	o.write("rm /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/4-prokka/result/"+file.split(".f")[0]+".ffn\n")
	o.write("rm /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/4-prokka/result/"+file.split(".f")[0]+".fna\n")
	o.write("rm /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/4-prokka/result/"+file.split(".f")[0]+".fsa\n")
	o.write("rm /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/4-prokka/result/"+file.split(".f")[0]+".gbk\n")
	o.write("rm /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/4-prokka/result/"+file.split(".f")[0]+".sqn\n")
	o.write("rm /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/4-prokka/result/"+file.split(".f")[0]+".tbl\n")
```

```shell
split -l 3753 -d prokka.sh prokka_split_
file in prokka_split_*; do mv $file "$file.sh"; done
for file in prokka_split_*; do nohup bash $file > "$file.log" 2>&1 & done
```

```shell
conda create -n roary
conda activate roary
conda config --add channels r
conda config --add channels defaults
conda config --add channels conda-forge
conda config --add channels bioconda
conda install roary

nohup roary -v -p 96 /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/4-prokka/result/*.gff -f /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/5-roary/ &
```

提取核心基因组的序列： [核心基因提取：prokka、roary和TBtools的联合运用 - 简书 (jianshu.com)](https://www.jianshu.com/p/f99e62d08ac9) 

## 19.mummer call snp

```python
import os
o=open("/workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/6-mummer/mummer.sh","w")
path=["/workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/1-not_301_strain/1-971_strain/","/workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/2-1285_contig/","/workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/3-5391_strain/"]
for p in path:
	files=os.listdir(p)
	for file in files:
		prefix=file.split(".f")[0]
		o.write("nucmer --mum -p "+prefix+" /workdir/Liuxq/proj/20230423_Pseudomonas_aeruginosa/0-reference/GCF_000006765.1_ASM676v1_genomic.fna "+p+file+"\n")
		o.write("delta-filter -1 -q -r "+prefix+".delta > "+prefix+".filter\n")
		o.write("show-snps -C -T -r -l "+prefix+".filter > "+prefix+".vcf\n")
```

## 20.建树

```shell
核心基因snp建树：2097-大树(2097个菌株，包括reference GCA_002813595.1)
cd /workdir/lishsh/2-Hvkp/6-TREE/Total-Strain/Total-Tree-roud7
cat /workdir/lishsh/2-Hvkp/6-TREE/Total-Strain/Total-Tree-roud7/total-snp/* >total.snp
cat /workdir/lishsh/2-Hvkp/6-TREE/Total-Strain/Total-Tree-roud7/total-indel/* >total.indel

cut -f "1,2" total.snp > total_snp_site.txt
cut -f "1" total.indel > total_indel_site.txt
sort -n total_snp_site.txt | uniq > total_snp_uniq.txt
sort -n total_indel_site.txt | uniq > total_indel_uniq.txt

perl 6-snp_site_filter_indel.pl
得到：total_snp_noindel.txt
440998 total_snp_noindel.txt
perl 6-remove_mobile.pl
424980 total_snp_noindel_norepeat.txt
6.对删除重复区、可移动元件区的snp位点排序
sort -n total_snp_noindel_norepeat.txt > sort_total_snp_noindel_norepeat.txt
424980 sort_total_snp_noindel_norepeat.txt
根据泛基因组结果挑选核心基因（注意跑roary的时候把reference加上），按照99%挑选核心基因，本次2193个
整理核心基因位点在reference上面对应的位置，两列：roary_core_gene_site.txt
perl select_core_gene_snp_site.pl ##整理出只在roary_core_gene_site.txt位点的snp，作为核心的snp集
139122 total_core_snp_noindel_norepeat.txt
7.将每个菌株的snp文件生成 
perl 9-SNP_sort_all.pl  将每个菌株的snp 文件存成hash， 读取sort_total_snp_noindel_norepeat.txt 如果存在snp，就用自己的碱基，没有snp就用reference的碱基
8.将snp文件转为fasta格式
perl 10-SNP_sort_seqs.pl

cd /workdir/lishsh/2-Hvkp/6-TREE/Total-Strain/Total-Tree-roud7/outgroup
11.选择外群（先call snp）
perl 8-outgroup_snp_sort.pl 
12.将外群的snp文件转化为fasta文件
perl 9-outgroup_snp_seq.pl
13.将所有菌株的序列与外群的序列连接在一起

14.利用去重组前的139122个SNP建树
cd /workdir/lishsh/2-Hvkp/6-TREE/Total-Strain/Total-Tree-roud7/outgroup-tree
cat *.fasta >total-and-outgroup.fasta

nohup iqtree -s 2097-and-outgroup.fasta -B 1000 -m GTR+I+G -T 50 > ./iqtree.log 2>&1 &
nohup iqtree -s 2097.fasta -B 1000 -m GTR+I+G -T 50 > ./iqtree.log 2>&1 &

15.去重组，计算重组率：
ClonalFrameML 2097.fasta.treefile 2097-whole-snp.fasta 2097_clonal -em true -emsim 100 

整理ClonalFrameML得到的结果后，去掉重组区域的snp
perl 6-remove_chongzu.pl
对139122个snp去重组后剩余65713个snp
65713 last_quzhongzu_core_snp_noindel_norepeat.txt

16.整理snp后建树

将每个菌株的snp文件生成 
perl 9-SNP_sort_all.pl  将每个菌株的snp 文件存成hash， 读取sort_total_snp_noindel_norepeat.txt 如果存在snp，就用自己的碱基，没有snp就用reference的碱基
17.将snp文件转为fasta格式
perl 10-SNP_sort_seqs.pl

cd /workdir/lishsh/2-Hvkp/6-TREE/Total-Strain/Total-Tree-roud7/tree-after-quzhongzu
cat ./snp-fasta/* >2097-quchongzu.fasta
18.建树
cd /workdir/lishsh/2-Hvkp/6-TREE/Total-Strain/Total-Tree-roud7/tree-after-quzhongzu/outgroup-tree
cat 2097-quchongzu.fasta   outgroup.fasta >2097-quchongzu-and-outgroup.fasta
nohup iqtree -s 2097-quchongzu-and-outgroup.fasta -B 1000 -m GTR+I+G -T 45 > ./iqtree.log 2>&1 &

cd /workdir/lishsh/2-Hvkp/6-TREE/Total-Strain/Total-Tree-roud7/tree-after-quzhongzu/no-outgroup-tree
nohup iqtree -s 2097-quchongzu.fasta -B 1000 -m GTR+I+G -T 50 > ./iqtree.log 2>&1 &

export OMP_NUM_THREADS=16
nohup FastTreeMP -nt -gtr -n 1000 8-7647_outgroup.fasta > ./fasttree/7647_outgroup.fasttree &
nohup FastTreeMP -nt -gtr -n 1000 7-7647.fasta > ./fasttree/7647.fasttree &
```

## 21.cog分析

```R
setwd("C:/Users/liuxi/Desktop")
library(ggplot2)
library(dplyr)
# 构造数据
data = read.table("cog.txt",header = T,sep = "\t")
result <- data %>%
  arrange(class, percent)
result$cog=factor(result$cog,levels = result$cog)
p=ggplot(result, aes(x = percent, y = cog, fill = I("#456990"))) +
  geom_bar(stat = "identity", width = 0.6) +
  scale_fill_identity() +
  labs(x = "Proportion of core genes(%)", y = "COG_category") +
  theme_bw()+
  theme(text = element_text(color = "black"),
        axis.title.x = element_text(size = 15), 
        axis.title.y = element_text(size = 15),
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12))
pdf("core_genes_annotation.pdf",width = 8,height = 5)
p
dev.off()
```

## 22.碳青霉烯耐药机制

```R
library(tidyverse)
library(readxl)
setwd("C:/Users/liuxi/Desktop")
world.dat<-map_data("world")

ggplot() +
  geom_polygon(data=world.dat,aes(x=long,y=lat,group=group),
               fill="#dedede")+
  theme_bw()+
  scale_y_continuous(expand = expansion(mult=c(0,0)))+
  scale_x_continuous(expand = expansion(add=c(0,0)))+
  theme(panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())+
  labs(x=NULL,y=NULL)-> world.map

pdf("world-map.pdf",width=10,height=6)
world.map
dev.off()

df=read.table("africa.txt",header = T,sep = "\t")
df$cr=factor(df$cr,levels = c("2 carbapenemase genes","blaVIM","blaIMP","blaKPC","blaNDM","blaGES","blaSPM","blaDIM","Other carbapenemase gene","None"))
library(ggpubr)
library(ggsci)
labs=paste0(df$number," (",round(df$number*100/sum(df$number),digits = 1),"%)")
p=ggdonutchart(df, "number",
             label = labs,
             lab.pos = "out",
             fill = "cr",                            
             color = "black",                            
             palette = rev(c(pal_npg("nrc",alpha = 0.8)(9),"#808180CC"))
)+theme(legend.position = "right")
pdf("legend.pdf")
p
dev.off()
```

```R
# 弦图
library(circlize)
library(ggsci)
setwd("C:/Users/liuxi/Desktop")
df = read.table("CC-carbapenemase.txt",sep = "\t",header = T,row.names = 1)
data = as.matrix(df)
grid.col = NULL
grid.col[rownames(data)] = pal_npg("nrc",alpha = 1)(9)
grid.col[colnames(data)] = c(pal_simpsons("springfield", alpha = 1)(10),"#D0DFE6")
pdf("CC-carbapenemase-chord.pdf")
chordDiagram(data,grid.col = grid.col,transparency=0.5)
dev.off()

```

